README
NAME
    Catalyst::Plugin::Unicode::Encoding - Unicode aware Catalyst

SYNOPSIS
        use Catalyst qw[Unicode::Encoding];

        MyApp->config( encoding => 'UTF-8' ); # A valid Encode encoding

DESCRIPTION
    On request, decodes all params from encoding into a sequence of logical
    characters. On response, encodes body into encoding.

METHODS
    encoding
        Returns an instance of an "Encode" encoding

            print $c->encoding->name

OVERLOADED METHODS
    finalize_headers
        Encodes body into encoding.

    prepare_uploads
        Decodes parameters, query_parameters, body_parameters and filenames
        in file uploads into a sequence of logical characters.

    prepare_action
        Decodes request arguments (i.e. "$c->request->arguments") and
        captures (i.e. "$c->request->captures").

    setup
        Setups "$c->encoding" with encoding specified in
        "$c->config->{encoding}".

    handle_unicode_encoding_exception ($exception_context)
        Method called when decoding process for a request fails.

        An $exception_context hashref is provided to allow you to override
        the behaviour of your application when given data with incorrect
        encodings.

        The default method throws exceptions in the case of invalid request
        parameters (resulting in a 500 error), but ignores errors in upload
        filenames.

        The keys passed in the $exception_context hash are:

        param_value
            The value which was not able to be decoded.

        error_msg
            The exception recieved from Encode.

        encoding_step
            What type of data was being decoded. Valid values are
            (currently) "params" - for request parameters / arguments /
            captures and "uploads" - for request upload filenames.

SEE ALSO
    Encode, Encode::Encoding, Catalyst::Plugin::Unicode, Catalyst.

AUTHORS
    Christian Hansen, "ch@ngmedia.com"

    Masahiro Chiba

    Tomas Doran, "bobtfish@bobtfish.net"

LICENSE
    This library is free software . You can redistribute it and/or modify it
    under the same terms as perl itself.



Changes
Revision history for Perl extension Catalyst::Plugin::Unicode::Encoding

1.9 2013-03-07 08:06:00
   - Bump Encode dependency, to avoid possibility of memory leak DoS.

1.8 2012-07-18 20:28:00
   - Work correcly with Shift_JIS and other encodings whos names are not
     upper case. RT#69677

1.7 2012-05-09 22:13:00
   - Add proper tests for charset parsing issues.
   - Fix all known charset parsing issues.

1.6 2012-05-05 14:18:00
    - And another issue with the same regex. RT#75006

1.5 2012-03-15 18:32:00
    - Fix issue with regex in last release. RT#75006

1.4 2012-03-13 22:28:00
    - Get charset parsing more correct in less trivial cases. RT#75006

1.3 2011-11-22 19:38:00
    - Correctly decode nested parameters (from Catalyst::Plugin::Params::Nested)
    - Require Class::Data::Inheritable. RT#71674
    - Require newer LWP version to avoid versions with unicode bugs.
      RT#64427

1.2 2011-09-07 21:19:00
    - Use a callback in failed decoding process so that the app has a chance
      to handle it nicely

1.1 2010-07-05 17:52:00
    - Skip Mech tests unless TWMC is a high enough version. (RT#55302)
    - Move Pod tests to only run for module authors
    - Do encoding slightly later so that we can catch the output of 
      finalize_error. Encoding now happens before finalize_headers rather
      than before finalize.

1.0 2010-05-04 21:46:00
    - Fix version number to be higher.

0.10 2010-05-04 21:30:00
	- Correctly decode UTF-8 URI Encoded (as per RFC3986) into
          $c->request->captures (wreis)

0.9 2010-03-07 22:50:00
        - Correctly decode UTF-8 URI Encoded (as per RFC3986) into
          $c->request->arguments
        - Test for this (u-foka)

0.8 2010-01-11 13:36:16
        - Don't fail tests if we have an old version of URI::Escape
          installed.

0.7 2009-12-21 21:30:12
        - Make Test::WWW::Mechanize::Catalyst an optional dependency.

0.6 2009-11-16 20:45:43
        - Fix missing dependency for the tests.
        - Avoid LWP releases which break the tests by messing up unicode.

0.5 2009-11-05 02:03:44
        - Decode $c->req->query_parameters, $c->req->body_parameters
        - Decode file upload filenames in parameters
        - Decode filenames in file upload objects
        - Don't die if content already in character octets is encountered

0.4  2009-10-16 17:07:07
        - Fix bug causing Encode to segfault in some cases
        - Fix broken tests

0.3  2009-05-28 19:01:15
        - Switch to Module::Install
        - Require Catalyst 5.8
        - Allow setting undef encoding (per request) to disable default of
          UTF-8
        - Don't overwrite encoding set in content types: 
            "text/html; charset=utf-8"
        - Handles the latin-1 (i.e. e-accute) case of encoding that everyone
          else forgets
        - Copy live tests from Catalyst::Plugin::Encoding
        - Fix bug with setting encoding in config (test from abraxxa)

0.2  2009-04-28 11:43:00 
        - Switch to MRO::Compat for Catalyst 5.8

0.1  2005-06-17 23:29:00
        - inital CPAN release.


t/lib/TestApp.pm
package TestApp;
use strict;
use warnings;
use base qw/Catalyst/;
use Catalyst qw/Unicode::Encoding/;

__PACKAGE__->config(
  encoding => $ENV{TESTAPP_ENCODING}
) if $ENV{TESTAPP_ENCODING};

__PACKAGE__->config('name' => 'TestApp');

__PACKAGE__->setup;

sub handle_unicode_encoding_exception {
  my ( $self, $param_value, $error_msg ) = @_;
  return $param_value;
}

1;


t/lib/TestApp2.pm
package TestApp2;
use strict;
use warnings;
use base qw/Catalyst/;
use Catalyst qw/Unicode::Encoding Params::Nested/;

__PACKAGE__->config(
  encoding => $ENV{TESTAPP_ENCODING}
) if $ENV{TESTAPP_ENCODING};

__PACKAGE__->config('name' => 'TestApp2');

__PACKAGE__->setup;

sub handle_unicode_encoding_exception {
  my ( $self, $param_value, $error_msg ) = @_;
  return $param_value;
}

1;


t/lib/TestLogger.pm
package TestLogger;
use strict;
use warnings;

our @LOGS;

sub new {
    return bless {}, __PACKAGE__;
}

sub debug {
    shift;
    push(@LOGS, shift());
}

1;



Makefile.PL
use inc::Module::Install 0.91;
use Module::Install::AuthorRequires;
use Module::Install::AuthorTests;

all_from 'lib/Catalyst/Plugin/Unicode/Encoding.pm';

requires 'Class::Data::Inheritable';
requires 'Encode' => '2.49';
requires 'Catalyst' => '5.90015'; # Will work with earlier, but 5.90014 had a test failing bug..
requires 'LWP' => '5.837'; # LWP had unicode fail in 5.8.26
requires 'URI' => '1.36';
requires 'Try::Tiny';

author_requires 'Catalyst::Plugin::Params::Nested';
author_requires 'Test::WWW::Mechanize::Catalyst' => 0.51;
test_requires 'Test::More' => '0.88';
test_requires 'IO::Scalar';
test_requires 'HTTP::Status';

author_tests 't/author';

resources repository => 'http://dev.catalyst.perl.org/repos/Catalyst/Catalyst-Plugin-Unicode-Encoding/';

if ($Module::Install::AUTHOR) {
    system("pod2text lib/Catalyst/Plugin/Unicode/Encoding.pm > README") and die;
}

WriteAll;



lib/Catalyst/Plugin/Unicode/Encoding.pm
package Catalyst::Plugin::Unicode::Encoding;

use strict;
use base 'Class::Data::Inheritable';

use Carp ();
use MRO::Compat;
use Try::Tiny;

use Encode 2.21 ();
our $CHECK = Encode::FB_CROAK | Encode::LEAVE_SRC;

our $VERSION = '1.9';

__PACKAGE__->mk_classdata('_encoding');

sub encoding {
    my $c = shift;
    my $encoding;

    if ( scalar @_ ) {
        # Let it be set to undef
        if (my $wanted = shift)  {
            $encoding = Encode::find_encoding($wanted)
              or Carp::croak( qq/Unknown encoding '$wanted'/ );
        }

        $encoding = ref $c
                  ? $c->{encoding} = $encoding
                  : $c->_encoding($encoding);
    } else {
      $encoding = ref $c && exists $c->{encoding}
                ? $c->{encoding}
                : $c->_encoding;
    }

    return $encoding;
}

sub finalize_headers {
    my $c = shift;

    my $body = $c->response->body;

    return $c->next::method(@_)
      unless defined($body);

    my $enc = $c->encoding;

    return $c->next::method(@_)
      unless $enc;

    my ($ct, $ct_enc) = $c->response->content_type;

    # Only touch 'text-like' contents
    return $c->next::method(@_)
      unless $c->response->content_type =~ /^text|xml$|javascript$/;

    if ($ct_enc && $ct_enc =~ /charset=([^;]*)/) {
        if (uc($1) ne uc($enc->mime_name)) {
            $c->log->debug("Unicode::Encoding is set to encode in '" .
                           $enc->mime_name .
                           "', content type is '$1', not encoding ");
            return $c->next::method(@_);
        }
    } else {
        $c->res->content_type($c->res->content_type . "; charset=" . $enc->mime_name);
    }

    # Encode expects plain scalars (IV, NV or PV) and segfaults on ref's
    $c->response->body( $c->encoding->encode( $body, $CHECK ) )
        if ref(\$body) eq 'SCALAR';

    $c->next::method(@_);
}

# Note we have to hook here as uploads also add to the request parameters
sub prepare_uploads {
    my $c = shift;

    $c->next::method(@_);

    my $enc = $c->encoding;

    for my $key (qw/ parameters query_parameters body_parameters /) {
        for my $value ( values %{ $c->request->{$key} } ) {
            # N.B. Check if already a character string and if so do not try to double decode.
            #      http://www.mail-archive.com/catalyst@lists.scsys.co.uk/msg02350.html
            #      this avoids exception if we have already decoded content, and is _not_ the
            #      same as not encoding on output which is bad news (as it does the wrong thing
            #      for latin1 chars for example)..
            $value = $c->_handle_unicode_decoding($value);
        }
    }
    for my $value ( values %{ $c->request->uploads } ) {
        # skip if it fails for uploads, as we don't usually want uploads touched
        # in any way
        $_->{filename} = try {
        $enc->decode( $_->{filename}, $CHECK )
    } catch {
        $c->handle_unicode_encoding_exception({
            param_value => $_->{filename},
            error_msg => $_,
            encoding_step => 'uploads',
        });
    } for ( ref($value) eq 'ARRAY' ? @{$value} : $value );
    }
}

sub prepare_action {
    my $c = shift;

    my $ret = $c->next::method(@_);

    foreach (@{$c->req->arguments}, @{$c->req->captures}) {
      $_ = $c->_handle_param_unicode_decoding($_);
    }

    return $ret;
}

sub setup {
    my $self = shift;

    my $conf = $self->config;

    # Allow an explict undef encoding to disable default of utf-8
    my $enc = exists $conf->{encoding} ? delete $conf->{encoding} : 'UTF-8';
    $self->encoding( $enc );

    return $self->next::method(@_);
}

sub _handle_unicode_decoding {
    my ( $self, $value ) = @_;

    return unless defined $value;

    if ( ref $value eq 'ARRAY' ) {
        foreach ( @$value ) {
            $_ = $self->_handle_unicode_decoding($_);
        }
        return $value;
    }
    elsif ( ref $value eq 'HASH' ) {
        foreach ( values %$value ) {
            $_ = $self->_handle_unicode_decoding($_);
        }
        return $value;
    }
    else {
        return $self->_handle_param_unicode_decoding($value);
    }
}

sub _handle_param_unicode_decoding {
    my ( $self, $value ) = @_;
    my $enc = $self->encoding;
    return try {
        Encode::is_utf8( $value ) ?
            $value
        : $enc->decode( $value, $CHECK );
    }
    catch {
        $self->handle_unicode_encoding_exception({
            param_value => $value,
            error_msg => $_,
            encoding_step => 'params',
        });
    };
}

sub handle_unicode_encoding_exception {
    my ( $self, $exception_ctx ) = @_;
    die $exception_ctx->{error_msg};
}

1;

__END__

=head1 NAME

Catalyst::Plugin::Unicode::Encoding - Unicode aware Catalyst

=head1 SYNOPSIS

    use Catalyst qw[Unicode::Encoding];

    MyApp->config( encoding => 'UTF-8' ); # A valid Encode encoding


=head1 DESCRIPTION

On request, decodes all params from encoding into a sequence of
logical characters. On response, encodes body into encoding.

=head1 METHODS

=over 4

=item encoding

Returns an instance of an C<Encode> encoding

    print $c->encoding->name

=back

=head1 OVERLOADED METHODS

=over

=item finalize_headers

Encodes body into encoding.

=item prepare_uploads

Decodes parameters, query_parameters, body_parameters and filenames
in file uploads into a sequence of logical characters.

=item prepare_action

Decodes request arguments (i.e. C<< $c->request->arguments >>) and
captures (i.e. C<< $c->request->captures >>).

=item setup

Setups C<< $c->encoding >> with encoding specified in C<< $c->config->{encoding} >>.

=item handle_unicode_encoding_exception ($exception_context)

Method called when decoding process for a request fails.

An C<$exception_context> hashref is provided to allow you to override the
behaviour of your application when given data with incorrect encodings.

The default method throws exceptions in the case of invalid request parameters
(resulting in a 500 error), but ignores errors in upload filenames.

The keys passed in the C<$exception_context> hash are:

=over

=item param_value

The value which was not able to be decoded.

=item error_msg

The exception recieved from L<Encode>.

=item encoding_step

What type of data was being decoded. Valid values are (currently)
C<params> - for request parameters / arguments / captures
and C<uploads> - for request upload filenames.

=back

=back

=head1 SEE ALSO

L<Encode>, L<Encode::Encoding>, L<Catalyst::Plugin::Unicode>, L<Catalyst>.

=head1 AUTHORS

Christian Hansen, C<ch@ngmedia.com>

Masahiro Chiba

Tomas Doran, C<bobtfish@bobtfish.net>

=head1 LICENSE

This library is free software . You can redistribute it and/or modify
it under the same terms as perl itself.

=cut
